File: examples/validation_error_structure.go
// Package main demonstrates how to inspect protovalidate ValidationError structure
package main

import (
	"fmt"

	"buf.build/go/protovalidate"
	userv1 "github.com/poi2/protovalidate-poc/go/gen/user/v1"
)

func main() {
	validator, err := protovalidate.New()
	if err != nil {
		panic(err)
	}

	examples := []struct {
		name    string
		request *userv1.CreateUserRequest
	}{
		{
			name: "Password mismatch",
			request: &userv1.CreateUserRequest{
				Name:                 "John Doe",
				Email:                "john@example.com",
				Password:             "password123",
				PasswordConfirmation: "different",
			},
		},
		{
			name: "Multiple field errors",
			request: &userv1.CreateUserRequest{
				Name:                 "", // empty name
				Email:                "not-an-email", // invalid email
				Password:             "short", // too short
				PasswordConfirmation: "short",
			},
		},
	}

	for _, ex := range examples {
		fmt.Printf("=== %s ===\n\n", ex.name)
		testValidation(validator, ex.request)
		fmt.Println()
	}
}

func testValidation(validator protovalidate.Validator, req *userv1.CreateUserRequest) {
	err := validator.Validate(req)
	if err == nil {
		fmt.Println("No validation errors")
		return
	}

	fmt.Printf("Error type: %T\n", err)
	fmt.Printf("Error: %v\n\n", err)

	if valErr, ok := err.(*protovalidate.ValidationError); ok {
		fmt.Printf("Number of violations: %d\n\n", len(valErr.Violations))
		for i, violation := range valErr.Violations {
			fmt.Printf("Violation %d:\n", i+1)
			fmt.Printf("  String: %s\n", violation.String())
			if violation.Proto != nil {
				if violation.Proto.Field != nil {
					fmt.Printf("  Field: %v\n", violation.Proto.Field)
				}
				if violation.Proto.Rule != nil {
					fmt.Printf("  Rule: %v\n", violation.Proto.Rule)
				}
				if violation.Proto.RuleId != nil {
					fmt.Printf("  RuleId: %s\n", *violation.Proto.RuleId)
				}
				if violation.Proto.Message != nil {
					fmt.Printf("  Message: %s\n", *violation.Proto.Message)
				}
				if violation.Proto.ForKey != nil {
					fmt.Printf("  ForKey: %v\n", *violation.Proto.ForKey)
				}
			}
			fmt.Println()
		}
	}
}

File: ts/examples/validationErrorStructure.ts
/**
 * Example: ValidationError structure
 * ValidationErrorの構造とviolationsの内容を実演
 */

import { create } from '@bufbuild/protobuf';
import { ProtoValidator, ValidationError } from '../src/validation/validator';
import { CreateUserRequestSchema } from '@gen/user/v1/user_pb';

console.log('=== ValidationError Structure Example ===\n');

const validator = new ProtoValidator();

// Create an invalid request with multiple validation errors
const invalidRequest = create(CreateUserRequestSchema, {
  name: '',                          // Too short
  email: 'not-an-email',             // Invalid email
  password: 'password123',
  passwordConfirmation: 'different', // Password mismatch
});

console.log('Invalid request:');
console.log(JSON.stringify(invalidRequest, null, 2));
console.log();

try {
  validator.validate(CreateUserRequestSchema, invalidRequest);
  console.log('No validation errors (unexpected)');
} catch (error) {
  if (error instanceof ValidationError) {
    console.log('ValidationError caught!');
    console.log(`\nError message: ${error.message}`);
    console.log(`\nNumber of violations: ${error.violations.length}`);
    console.log('\nViolation details:');

    error.violations.forEach((violation, index) => {
      console.log(`\n  Violation ${index + 1}:`);
      console.log(`    Field Path:  ${violation.fieldPath}`);
      console.log(`    Message:     ${violation.message}`);
      console.log(`    Rule ID:     ${violation.ruleId}`);
      console.log(`    Reason Code: ${violation.reasonCode}`);
    });

    console.log('\n=== JSON representation ===');
    console.log(JSON.stringify(error.violations, null, 2));
  } else {
    console.error('Unexpected error:', error);
  }
}
